engine/kv_engine.cpp

```diff
@@ -1496,15 +1512,169 @@ Status KVEngine::SGet(const StringView collection, const StringView user_key,
+                                ExpiredTimeType* expired_time) {
+  HashTable::KeyHashHint hint = hash_table_->GetHint(str);
+  HashEntry hash_entry;
+  HashEntry* entry_ptr = nullptr;
+  *expired_time = kInvalidExpiredTime;
+  Status s = hash_table_->SearchForRead(hint, str, ExpiredRecordType,
+                                        &entry_ptr, &hash_entry, nullptr);
+  if (s != Status::Ok) {
+    return s;
+  }
+
+  Collection* collection_ptr = nullptr;
+  switch (entry_ptr->GetIndexType()) {
+    case HashIndexType::Skiplist: {
+      collection_ptr = entry_ptr->GetIndex().skiplist;
+      *expired_time = collection_ptr->GetExpiredTime();
+      break;
+    }
+    case HashIndexType::UnorderedCollection: {
+      collection_ptr = entry_ptr->GetIndex().p_unordered_collection;
+      *expired_time = collection_ptr->GetExpiredTime();
+      break;
+    }
+
+    case HashIndexType::Queue: {
+      collection_ptr = entry_ptr->GetIndex().queue_ptr;
+      *expired_time = collection_ptr->GetExpiredTime();
+      break;
+    }
+
+    case HashIndexType::StringRecord:
+      *expired_time = entry_ptr->GetIndex().string_record->GetExpiredTime();
+      break;
+    default:
+      return Status::NotSupported;
+  }
+  // TODO(zhichen): Free this record
+  if (*expired_time > 0 && *expired_time <= TimeUtils::millisecond_time()) {
+    if (collection_ptr) {
+      // set hash entry empty
+      hash_table_->Erase(entry_ptr);
+      // TODO(zhichen): push background thread cleaner
+    }
+    *expired_time = kInvalidExpiredTime;
+    return Status::NotFound;
+  }
+  // convert to millisecond
+  return Status::Ok;
+}
+
+Status KVEngine::UpdateHeadWithExpiredTime(Skiplist* skiplist,
+                                           ExpiredTimeType expired_time) {
+  auto value = skiplist->header()->record->Value();
+  auto key = skiplist->header()->record->Key();
+  auto request_size = value.size() + key.size() + sizeof(DLRecord);
+  SpaceEntry sized_space_entry = pmem_allocator_->Allocate(request_size);
+  if (sized_space_entry.size == 0) {
+    return Status::PmemOverflow;
+  }
+  void* new_head_record =
+      pmem_allocator_->offset2addr(sized_space_entry.offset);
+
+  while (1) {
+    SkiplistNode* dram_node = nullptr;
+    HashEntry* entry_ptr = nullptr;
+    HashEntry hash_entry;
+    DataEntry data_entry;
+    DLRecord* existing_record = nullptr;
+    auto hint = hash_table_->GetHint(key);
+    std::unique_lock<SpinMutex> ul(*hint.spin);
+    version_controller_.HoldLocalSnapshot();
+    defer(version_controller_.ReleaseLocalSnapshot());
+    TimeStampType new_ts =
+        version_controller_.GetLocalSnapshot().GetTimestamp();
+    Status s = hash_table_->SearchForRead(hint, key, SortedHeaderRecord,
+                                          &entry_ptr, &hash_entry, &data_entry);
+    if (s != Status::Ok) return s;
+    dram_node = entry_ptr->GetIndex().skiplist->header();
+    existing_record = dram_node->record;
```

Comment:
SkiplistHead is not indexed by SkiplistNode. In HashTable, there is a Skiplist* stored in HashEntry. You should access the underlying DLRecord in that Skiplist object.